File: matlab/RLSFinal.m
%RLS beamsteering algorithm
%Jake Hogan
function[weights,error] = RLS(lambda,delta,sysorder)
%% Recursive Least Squares
% Input arguments: 
%   lambda = forgetting factor, dim 1x1 
%   sysorder = filter order, dim 1x1 
%   received = input signal, dim Nx1 
%   desired = desired signal, dim Nx1
%   delta = initial value, P(0)=delta^-1*(Identity Matrix), dim 1x1 
% Output arguments:
%   error = a priori estimation error, dim nx1
%   weights = final filter coefficients, dim sysorderx1

%% create signal    
    % Data Parameters
    numPts  = 1000;             % number of points to generate
    freq    = 12e9;              % frequency of fundamental tone
    filtord = 16;                % filter order
    filt    = rand(filtord, 1); % filter coefficients
    nVar    = 1;                % white noise variance
    SNR     = -20;              % signal to noise ratio of tone
    
    [received,desired,noise] = genSignal(numPts, freq, filt, nVar, SNR);
    
%% intitalize algorithm
%lambda = ;
%sysorder = ;
%delta = ;
P=eye(sysorder)* 1/delta; % identity matrix with delta as diagonals
weights=zeros(sysorder,1); % zeroing out initial weights
received = received(:);% makes the signal column vectors
error = desired*0;
lambda1 = (lambda)^-1;% inverse of forgetting factor
len = length(received);

%% begin
for n = sysorder:len
    input = noise(n:-1:n-sysorder+1);% corrects dimensions

    K =(lambda1*P*input)/(1+lambda1*input'*P*input); % calculates Gain matrix
    
    output = weights' * input ; % calculates array output w/ ccurrent weights
    
    error(n) = received(n)-output; % a priori error of the signal
       
    weights = weights + K*error(n); % updates weights
    
    P=(lambda1*P)-(lambda1*K*input'*P); % calculates new correlation matrix
    
    
end

e = error(16:1000);
q = linspace(1,985,985);

e2 = abs(e.^2);



%--------------------------------------------------------------------------
% Plot
%--------------------------------------------------------------------------

% Plot filter results
t = linspace(0,length(received)/numPts,length(received));
%figure;
%plot(t,received,t,error);

%title('Comparison of Filtered Signal to Reference Input');


% Plot comparison of results to original signal 
    figure;
    plot(t,desired,t,error);
   
    title('Comparison of Filtered Signal to Original Signal');
%plot learning curve
figure;
plot(q, e2);





% Calculate SNR improvement
SNRi    = 10*log10(var(received)/var(error));

disp([num2str(SNRi) 'dB SNR Improvement'])



return

function[received,desired,noise] = genSignal(numPts, freq, filt, nVar, SNR)

    % Generate time values
    t = linspace(0,1,numPts)';
    
    
    % Generate tone
    desired = sin(2*pi*freq*t);
    
    %wavevector from angle of arrival
    elevation = pi/3;
    azimuth = pi/3;
    g = [];
    for m=1:4
        u = exp(1i*pi*(m-1));
        a = u*sin(elevation)*cos(azimuth);
        g = [g a];
    end
    %for y elements
    h = [];
    for p=1:4
        v = exp(1i*pi*(p-1));
        b = v * sin(elevation)* sin(azimuth);
        h = [h b];
    end
    
    % Generate noise
    noise = sqrt(nVar)*randn(numPts,1);
    addnoise = filter(filt, 1, noise);
    freqz(filt,1,numPts)
    
    % Adjust SNR of tone
    desired = (desired)/sqrt(var(desired)/(10^(SNR/10)*var(noise)));
    disp(['Calculated SNR = ' num2str(10*log10(var(desired)/var(noise)))])
    
    % Add noise to signal
    received = (desired*g*h.') + addnoise ;
    
return

File: source/RLSFinal.cpp
/*
    UMass Lowell SPACEHAUC Beam Steering
    RLSFinal.cpp     
    GNU GPLv3 License
*/

#include <iostream>
#include <armadillo>
#include <cmath>
#include <vector>
#include <complex>
#include "sig/sigpack/sigpack.h"
#include <cstdlib>
#include <time.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

using namespace std;
using namespace arma;
using namespace sp;

typedef complex<double> cdouble;

class GenSignalReturn {
	/*Matrices returned by the genSignal function.
	These matrices are named exactly as found in the
	Matlab program.*/
public:
  mat received;
  mat desired;
  mat noise;

  GenSignalReturn(mat received, mat desired, mat noise) {
	  /*Create return object from matrices.*/
    this->received = received;
    this->desired = desired;
    this->noise = noise;
  }
};

class RLSReturn {
	/*Matrices returned by the RLS() function. Matrices
	are named as they appear in the Matlab project.*/
public:
  mat error;
  mat weights;

  RLSReturn(mat error, mat weights) {
    this->error = error;
    this->weights = weights;
  }
};

double genRandAngle(){
	/*Generate a random angle between 0 and pi.*/
    double randNum = ((double) rand() / (RAND_MAX));
    return randNum * M_PI;
}

GenSignalReturn genSignal(double num_points, double frequency, mat filt,
			  double n_var, double SNR) {
    /*
      Produces object containing three members:
      received      input signal, dim 1x1
      desired       desired signal, dim 1x1
      noise         noise 
      
      Input:
      @param num_points       number of points to generate 
      @param frequency         frequency of fundamental tone
      @param filt         filter coefficients
      @param n_var         white noise variance
      @param SNR          signal to noise ratio of tone
    */

  cdouble i(0, 1);

  mat t = linspace(0, 1, num_points);

  mat desired = sin(2*M_PI*frequency*t);
 
  double elevation = genRandAngle();
  double azimuth   = genRandAngle();
  
  std::vector<complex<double> > g;
  for(double m = 0; m < 4; m++) {
    cdouble u = exp(i*M_PI*(m-1));
    cdouble a = u*sin(elevation)*cos(azimuth);
    g.push_back(a);
  }
  cx_rowvec g_arma = cx_rowvec(g);

  std::vector<complex<double> > h;
  for(double p = 0; p < 4; p++) {
    cdouble v = exp(i*M_PI*(p-1));
    cdouble b = v * sin(elevation)*sin(azimuth);
    h.push_back(b);
  }
  cx_rowvec h_arma = cx_rowvec(h);

  // Generate noise
  FIR_filt<double, double, double> G;
  G.set_coeffs(filt);
  mat noise = sqrt(n_var) * randn(num_points, 1);
  mat addnoise = G.filter(noise);
 
  desired = (desired) / sqrt(var(vectorise(desired))/(pow(10,(SNR/10))*var(vectorise(noise))));

  // Add noise to signal - beware of the real function!

  cx_mat result = desired*g_arma*h_arma.t();
  
  mat received = real(result) + addnoise;
  
  return GenSignalReturn(received, desired, noise);
}


RLSReturn RLS(double lambda, double delta, double sysorder) {
  /*
  Input:
  @param lambda           forgetting factor, dim 1x1   
  @param delta            initial value, P(0)=delta^(-1) * (Identity Matrix), dim 1x1
  @param sysorder         filter order, dim 1x1
     
  Output:
  Object containing members .error and .weights
  @error                  a priority estimation error, dim nx1
  @weights                final filter coefficients, dim sysorderx1
  */

  arma_rng::set_seed_random();
  
  // parameters!
  int num_points = 1000;
  double frequency = 12e9;
  int filter_order = 16;
  mat filter = randu<mat>(filter_order, 1);
  int n_var = 1;
  int SNR = -20;
  
  // Generate some sort of signal - wow!
  GenSignalReturn gsr = genSignal(num_points, frequency, filter, n_var, SNR);
  mat received = gsr.received;
  mat desired = gsr.desired;
  mat noise = gsr.noise;

  mat P = eye(sysorder, sysorder) * 1.0/delta;
  mat weights = zeros<mat>(sysorder,1);
  vec received_vec = vectorise(received);
  mat error = desired*0;
  double lambda1 = pow(lambda, -1);
  int len = received_vec.n_elem;

  // Big calculation here
  int number_of_loop_executions = 0;
  for(int n = sysorder-1; n < len; n++) {
    ++number_of_loop_executions;

    mat input = zeros<mat>(sysorder, 1);
    for(int i = n; i > n-sysorder; i--) {
      //cout << "i: " << i << endl;
      input(n - i) = noise(i);
    }

    mat K = (lambda1*P*input)/as_scalar(1+lambda1*input.t()*P*input); //ahh!
	  
    mat output = weights.t() * input; // there was another '* here!

    error(n) = received_vec(n) - as_scalar(output);

    weights = weights + K * error(n); //error(n) instead

    P=(lambda1*P)-(lambda1*K*input.t()*P); // another '* here
  }
  
  mat e = zeros<mat>(1000-16, 1);

  for(int index = 0; index < 1000-16; index++) {
    double tmp = error(15+index);
    e(index) = tmp;
  }

  mat q = linspace(1, 985, 985);

  mat e2 = abs(pow(e,2));

  // Calculate SNR improvement
  double SNRi = 10*log10(var(vectorise(received))/var(vectorise(error)));
  cout << SNRi << "dB SNR Improvement" << endl;

  return RLSReturn(error, weights);
}


int main(int argc, char* argv[]) {
  
  srand (time(NULL));
  // Driver program for RLS() function.
  RLS(0.98, 100.0, 16.0);
  return 0;
}

